reactioncommerce/reaction
# feat: add product and productVariant related GraphQL mutations

Resolves parts of #5529
Impact: **minor**  
Type: **feature|refactor**

## Issue
Our `product` (not `catalog`) admin server code is still fully `meteor`, including a few mutations which use `getGraphQLContextInMeteorMethod`, which we are deprecating.

## Solution
1. Add four new GraphQL mutations for product admin:
- `createProduct`
```
mutation {
  createProduct(input: { shopId: "{shopId}" }){
    product {
      _id
    }
  }
}
```

- `createProductVariant`
```
mutation {
  createProductVariant(input: { productId: "{parentProductId}", shopId: "shopId" }){
    variant {
      _id
    }
  }
}
```

- `cloneProducts`
```
mutation {
  cloneProducts(input: { productIds: ["{id1}"," {id2}"], shopId: "shopId" }){
    products {
      _id
    }
  }
}
```

- `cloneProductVariants`
```
mutation {
  cloneProductVariants(input: { variantIds: ["{id1}"," {id2}"], shopId: "shopId" }){
    variants {
      _id
    }
  }
}
```

2. Remove existing `products/createProduct`, `products/createVariant`, `products/cloneVariant` code, and all related code which is no longer needed / moved in the new mutations

3. Update client code to call new GraphQL mutations instead of meteor methods.
- This was a much bigger task than was expected. A lot of our client components use `UNSAFE_...` and other React lifecycle methods, and will need to be refactored completely in the near future, so I did my best to keep them as close to how they are as possible as to not need a complete refactor at the moment.

One regression / change here is that when you create a new product / variant, you are no longer redirect to that new variant page. The reason is when we use a GraphQL mutation, the new product / variant is created almost immediately, however, it takes 1+ seconds for the Meteor subscriptions to update and include the new product. Because of this, when redirecting, you'd land on the page of a variant that isn't yet in the Meteor subscription, and there would be an error. Because of the need to refactor the components soon anyway, I think this is a decent trade off, as opposed to delaying this PR and refactoring all of these components now. I've left a comment in the code suggesting we re-add this feature when these components are updated.

## Breaking changes
Any custom plugins which use any of the above listed meteor methods will need to be updated to use the GraphQL mutation instead


## Testing
1. Create a product using the mutation, and also in the UI (Create Product button above Product table)
1. Create a variant using the mutation, and also in the UI (+ button on product page)
1. Create an option using the mutation, and also in the UI (+ button on product variant page in options area)
1. Clone a product using the mutation, and also in the UI (check one or more products on the product table, and use the dropdown to clone, and also inside an individual product page, use the `...` menu)
1. Clone a variant using the mutation, and also in the UI (inside an individual product variant, use the `...` menu))
1. See all of the above work

You can see that all the new resolvers, `shop`, `tags`, 'slug'  and `variants` on type `Product`, and `options` and `shop` on `ProductVariant`, work with this mutation:
```
mutation {
  cloneProducts(input: { productIds: ["{productId}"], shopId: "{shopId}" }){
    products {
      _id,
     slug
      shop {
				_id
        name
      }
      tagIds,
      tags {
        edges {
          node {
            displayTitle
            heroMediaUrl
            position
            slug
          }
        }
      },
      variants {
        _id
        shop {
           _id
          }
        options {
          _id
         shop {
           _id
          }
        }
      }
    }
  }
}
```
